### ISSUE TITLE
archive storage

### ISSUE THREAD

**Issue by toxic0berliner** [blakeblackshear/frigate#3673]
I want to be able to move old recordings to my NAS so that I can have recent recordings on my local fast SSD but still keep older footage available on my NAS having terabytes of spinning rust available.

**Describe the solution you'd like**
A way to configure an archive storage (either a different path that I mount using NFS or smb or even directly the connection parameters for frigate to use SMB client himself)
+ A way to configure the retention individually for main and archive storage (7 days in main storage and 90 days in archive for example)

**Describe alternatives you've considered**
No real idea, except running frigate twice and keeping duplicate of the "main" storage.

**Additional context**
Having 2 level of storage is the key here, moving data from main to archive storage can be complicated within frigate, so a first version could be to add a "cache" storage that fully duplicate the main storage but allows to setup shorter retention.
This way I could set retention to 90 days on main storage and point it to a NFS mount and add a "cache" retention of 7 days in the local SSD under /tmp.
This way it could be less complicated to implement: any recording in the db is available in main storage (NFS) if it was not found in the "cache" (SSD or /tmp)

**Comment by rhatguy** [comment]
Just thinking...could something like this be created with UnionFS and a separate script that moved all files over X days old to the "archive" filesystem.  I believe UnionFS would allow frigate to still see both drives as a continuous filesystem regardless of which disk the files actually live on.

https://en.wikipedia.org/wiki/UnionFS

**Comment by NickM-27** [comment]
yeah, I don't think this is something that makes sense for frigate to integrate itself. For example, my Unraid NAS already does this as the default behavior. There are shares which consist of one or many harddrives, and all writes are done to a cache drive first and moved to the harddrive when full. Frigate just sees this as one big drive. 

We don't want frigate managing the native filesystems, we want one volume mount for recordings without caring what the actual file system is behind that.

**Comment by toxic0berliner** [comment]
Seems a bit of a hack but could be possible. Unionfs seems quite old though or at least esoteric...
Given frigate has to manage massive volume of data I would expect archiving or caching strategies to fit many people's requirements but I might be mistaken.

**Comment by toxic0berliner** [comment]
I get that it's tempting to rely on the intelligent FS we have nowadays. But let's be honest too it will probably not be fine grained with a retention duration set in frigate, even people using zfs and cache layers might want to remove the recordings before the cache is actually full to keep some for other uses...

Managing storage seems like a must for me for a recorder of my IP cameras. If this is not in the future of frigate I'll have to look someplace else. 

It's not directly linked to this feature but consider that most NVR implement a rotation of recordings when the drive is full as one of the first feature, so managing the storage does make sense for an NVR

**Comment by NickM-27** [comment]
This can also be done with zfs which is a modern file system: https://unix.stackexchange.com/questions/316422/ssd-as-cache-for-filesystem-e-g-on-hdd-or-any-other

Other options as well: https://blog.delouw.ch/2020/01/29/using-lvm-cache-for-storage-tiering/

**Comment by NickM-27** [comment]
I agree it would be useful, as I use it myself, but I don't think it's something that makes sense for frigate to re-implement as many solutions already exist and we have many users already doing this.

**Comment by NickM-27** [comment]
> It's not directly linked to this feature but consider that most NVR implement a rotation of recordings when the drive is full as one of the first feature, so managing the storage does make sense for an NVR

This is already an existing feature request that is going to be implemented in #994

**Comment by rhatguy** [comment]
@toxic0berliner I'm not disagreeing with your feature request, but I think where this discussion really begins to be helpful is when you talk about tiering off to a cloud provider (probably via object).  I suspect many people would be interested in keeping 7 days onsite, then tiering off to S3 or backblaze.  Technically this could be achieved with a similar UnionFS type approach.  I feel like this really boils down to whether the frigate developers think the app needs to be intelligent enough to tier data, or whether they are going to leave that up to lower layers.  My personal opinion would be that there are more beneficial features to implement than tiering storage when there are other ways to accomplish tiering.

**Comment by toxic0berliner** [comment]
Not disagreeing with you either 😀
Storage tiering makes sens I believe but I could fully understand other priorities might take precedence.
I have little doubt that it's too complicated for my level of dev anyway so I'll be patient 😜

**Comment by javydekoning** [comment]
I disagree that this should be left to the Filesystem because that limits you to a single FS (type).

It's very common for container orchestrators to have different storage classes.

For example:

"Hot" mounted at: /media/frigate/recordings/hot
"Cold" mounted at: /media/frigate/recordings/cold

This allows recent recordings to be local (block/nvme/ramdisk/whatever) while the rest is moved to remote NFS for example. The only thing frigate would have to do is mv the object.

Implementing s3 compatible storage at a later point in time would be a nice to have as well. That would allow users to keep recordings at the cloud provider of their choice (durable and cost effective).

**Comment by sstratoti** [comment]
Just sharing my experience/use case. I've taken the plunge with Frigate off of Blue Iris, and so far the experience has been excellent. Especially with the .12 release! But in my previous setup, Blue Iris allowed me to set a size limit or an age limit for a directory of storage. 

It also allowed me the options to either delete or move older files to another directory as it reached that limit. 

It would run a maintenance job every few minutes (let's say 10, not actually sure) to move/rotate the files out or delete them. 

The same settings could be set for the "other directory". Either age or max size. 

This allows you to manage the maximum space that the app should use so you can leave other space on your main drive for other applications, docker containers, files, swap, etc. 

It would also allow me to keep the most recent snapshots and videos on my SSD, and move older files that I'd like to keep, but probably not access, in a slower USB 3.0 SATA external drive that is much larger. 

I could move everything over to the SATA drive, but then I'll have like 990 GB of SSD drive that'll just be sitting there unused. I could copy/archive things myself using rsync or something, but then I'd lose access to them in the timeline once the main files get removed by the new .12 file rotating feature. 

Unlesssss...if I move the files myself, would it be possible to update the SQLite DB to point at the new location of the files so that they aren't lost in the UI timeline? Are those entries cleaned up somehow?

**Comment by rhatguy** [comment]
Not arguing whether this approach is the "best" way to achieve what you want, BUT...if you're willing to move the files yourself you can easily use unionfs today to combine /ssd_mountpoint and /hdd_mountpoint into a new mountpoint /frigate_storage.  Then you can move files in the background manually and still have everything show up to frigate.  Frigate doesn't know which underlying filesystem the files live on.  No need to update the database.  Something like the below should accomplish what you want.  Then you just need to move the files manually from /ssd_mountpoint to /hdd_mountpoint at whatever criteria you decide.

mount -t unionfs -o dirs=/ssd_montpoint:/hdd_mountpoint=ro none /frigate_storage

**Comment by NickM-27** [comment]
I run in unraid which does this automatically, it uses fuse.fs

**Comment by kravemir** [comment]
> I want to be able to move old recordings to my NAS so that I can have recent recordings on my local fast SSD but still keep older footage available on my NAS having terabytes of spinning rust available.

This is already a solution, which is to "move" recordings, and not a problem statement, of what do you want to solve in real life.

Real life problems statement might be:

- have recent recordings available as proof of a burglary or an incident:
  -  what's point of recording if not to have a proof?
  - in case of physical breach, the proof of burglary should not be stored only on just one local device, which increases probability of proof material being destroyed or stolen during burglary,
- have recent recordings available on SSD for fast browsing,
- have old recordings available on HDD on NAS for archivation:
  - cover very low probability of a need to dig into those in future.

However, when this is combined with other environmental constraints:

- solution is deployed in house/SOHO/facility with metered LTE or 5G connection, and it's not possible to constantly stream out camera recordings.

Depending on environmental constraints,

- one might stream out camera footage to remote frigate instance (via VPN),
- one might record simultaneously to multiple storage devices via single frigate instance:
  - keep duration is low for SSD device,
  - keep duration is high for HDD device,
- one might record simultaneously via multiple frigate instances:
  - primary instance has got all cool features (detection, browsing, events,..) enabled,
  - secondary (archival) instance has is only storing camera streams to HDD, which has low requirements on CPU/RAM.
  
In combination with above, when on doesn't do constant remote streaming/storing:
  
  - primary instance could be always sending only events out (JPEG, parts of stream,...) to remote storage,
  - primary instance could be sending only events in case it's turned on (i.e. away from house mode) to save data on metered connection - when people are present at site/home:
    - storage devices aren't unattended,
    - people generate much more events, and that increases amount of data sent to remote storage,
    - also these extra events aren't important to be saved on remote.
 
I wouldn't focus on how to solve things, as long as it solves real life problems in given environmental constrains, then I would just go with the easiest and/or cheapest available solution.
 
The constraint is something, that is not in power of owner to change, i.e. not that I want to do it this way. But, that in my location, I don't have any unmetered connection available.

**Comment by sstratoti** [comment]
My issue isn't that the files need to be moved. It's that once they're moved you can't access them within the UI to look for/at past events. 

When using unionfs or mergefs, my understanding is that it fills up one disk and then moves onto the next while combining it to look like a single directory. 

Doing it this way makes sense if you have multiple HDs that you want to combine into one, but if I want to utilize my SSD/NVMe drive for writes and then spinners for storage/archive, is there a way to do this unless Frigate controls which disk/folder it writes to and reads from?

If it sees it as a single directory, and FIFOs the files, would it maximize the benefit of the SSD?

I'm totally up for trying this method if my understanding of this is incorrect. Thanks!

**Comment by NickM-27** [comment]
> but if I want to utilize my SSD/NVMe drive for writes and then spinners for storage/archive, is there a way to do this unless Frigate controls which disk/folder it writes to and reads from?

Yes, this is exactly what unraid does via fuse.fs and it works perfectly. It appears as one volume to frigate, and the OS decides when to move files from SSD to the HDD archival storage

**Comment by sstratoti** [comment]
And this works if I'm not using unraid?

**Comment by NickM-27** [comment]
It's nothing proprietary to unraid, though unraid makes it really easy

**Comment by sstratoti** [comment]
Alright, I'll give it a try. Thank you!

**Comment by toxic0berliner** [comment]
Regardless, this is a feature request and the proposed "real life problem statement" doesn't match neither my needs nor that of others. Storage tiering is implemented in several NVR solutions and for good reasons, it can then be used in many ways to achieve whatever the user can think of building with it.

On my side I got it working with unions with a /hot that is being written to by default and a custom script+cron that moves files to /cold.

Nonetheless I feel this is not a proper solution and I still hope tiering will someday come to frigate. It has the DB and moving files shouldn't be too tricky, options for retention duration on each tear would benefit from being integrated as well as storage statistics. 
On my little ssd it took me a while to find how old files should be before I move them, doing the logic for moving files until drive is 80% full starting from the oldest files was too tricky for me to di in bash and not a long term solution I believe.

**Comment by javydekoning** [comment]
> > I want to be able to move old recordings to my NAS so that I can have recent recordings on my local fast SSD but still keep older footage available on my NAS having terabytes of spinning rust available.
> 
> 
> 
> This is already a solution, which is to "move" recordings, and not a problem statement, of what do you want to solve in real life.
> 
> 
> 
> Real life problems statement might be:
> 
> 
> 
> - have recent recordings available as proof of a burglary or an incident:
> 
>   -  what's point of recording if not to have a proof?
> 
>   - in case of physical breach, the proof of burglary should not be stored only on just one local device, which increases probability of proof material being destroyed or stolen during burglary,
> 
> - have recent recordings available on SSD for fast browsing,
> 
> - have old recordings available on HDD on NAS for archivation:
> 
>   - cover very low probability of a need to dig into those in future.
> 
> 
> 
> However, when this is combined with other environmental constraints:
> 
> 
> 
> - solution is deployed in house/SOHO/facility with metered LTE or 5G connection, and it's not possible to constantly stream out camera recordings.
> 
> 
> 
> Depending on environmental constraints,
> 
> 
> 
> - one might stream out camera footage to remote frigate instance (via VPN),
> 
> - one might record simultaneously to multiple storage devices via single frigate instance:
> 
>   - keep duration is low for SSD device,
> 
>   - keep duration is high for HDD device,
> 
> - one might record simultaneously via multiple frigate instances:
> 
>   - primary instance has got all cool features (detection, browsing, events,..) enabled,
> 
>   - secondary (archival) instance has is only storing camera streams to HDD, which has low requirements on CPU/RAM.
> 
>   
> 
> In combination with above, when on doesn't do constant remote streaming/storing:
> 
>   
> 
>   - primary instance could be always sending only events out (JPEG, parts of stream,...) to remote storage,
> 
>   - primary instance could be sending only events in case it's turned on (i.e. away from house mode) to save data on metered connection - when people are present at site/home:
> 
>     - storage devices aren't unattended,
> 
>     - people generate much more events, and that increases amount of data sent to remote storage,
> 
>     - also these extra events aren't important to be saved on remote.
> 
>  
> 
> I wouldn't focus on how to solve things, as long as it solves real life problems in given environmental constrains, then I would just go with the easiest and/or cheapest available solution.
> 
>  
> 
> The constraint is something, that is not in power of owner to change, i.e. not that I want to do it this way. But, that in my location, I don't have any unmetered connection available.
> 
> 

It's manual work. It's very typical to want to store your recordings on remote, durable storage.

Yes I can move it manually but then the Frigate interface won't be able to show them.

The s3 api is widely used and there are many providers offering affordable durable storage. It would be a good option to store recordingsAND metadata.

**Comment by kravemir** [comment]
> It's manual work. It's very typical to want to store your recordings on remote, durable storage.

It's typical to store recordings on remote store simultaneously while storing it locally.

This redundancy is comes useful especially when recording happens to be actually needed, such as burglary event, when the primary NVR might have been stolen in action, and therefore retain the ability to see recordings.

> ...  that I can have recent recordings on my local fast SSD but still keep older footage available on my NAS having terabytes of spinning rust available ...

For me, it makes no sense to implement such "move" functionality.

The most recent recording is often the most precious one. Storing it only on single non-redundant storage is increasing the risk of making security system useless - recordings stolen in action.

So, IMO, storing recordings simultaneously to multiple storages, where each storage has different retention period, is a better alternative, because it provides redundancy.

> Yes I can move it manually but then the Frigate interface won't be able to show them.
> 
> The s3 api is widely used and there are many providers offering affordable durable storage. It would be a good option to store recordingsAND metadata.

S3 based storage sounds as a viable option. The frigate would see it as one "logical" storage, while the physical storage would be redundant, where SSD and HDD is used for recent recordings, and also tiered, where old recordings deleted from SSD and kept on HDD. This would be transparent to frigate, and flexible - based on S3 storage configuration and capabilities.

The [minio](https://github.com/minio/minio) can be self-hosted, and there's something about [tiering](https://min.io/product/automated-data-tiering-lifecycle-management).

Though, not sure whether it would be possible to combine locally hosted S3 with some cloud provider, transparently, where local S3 storage, which is using local low-latency SSD, would work as LRU+write cache for provided remote S3 storage. So, maybe multi-storage support would be needed on frigate level.

**Comment by fpytloun** [comment]
I just migrated to Frigate from Zoneminder and really miss this feature. My use-case is that I have local storage of limited capacity and then Wasabi S3 mounted as POSIX FS via JuiceFS (k8s PVC). I would like to have option to move records older than x days into this cold storage with longer retention. With ZM it was done simply by having filter actions.

**Comment by fpytloun** [comment]
Looking at sqlite and it stores path so implementing move script would be simple. Basically just iterate over directories which are by date, move those matching threshold into another location and execute `UPDATE` statement to update path in sqlite.

**Comment by figadore** [comment]
>  this should work

This looks great, have you tested it?  This will still work with all of Frigate's "retain" logic, right? I'd love to see something similar as a built-in feature, but it sounds like this will do the trick until then.

**Comment by fpytloun** [comment]
> > this should work
> 
> This looks great, have you tested it? This will still work with all of Frigate's "retain" logic, right? I'd love to see something similar as a built-in feature, but it sounds like this will do the trick until then.

Yes, I am using it since sharing and it works nicely. And should work with retain logic as well :-)

**Comment by WilldabeastHA** [comment]
@fpytloun thank you for sharing this! I have a similar set up where i want to archive after a certain date to spinning disk. Forgive me for my lack of experience with scripts. How do I implement your above script into Frigate. I understand that I will mount /media/archive which will reside on the spinning disk but how do I automate the running of this script?

Thanks so much!

**Comment by fpytloun** [comment]
@WilldabeastHA If you are running in k8s, you can use this CronJob manifest and just tune that volumes part:
```
apiVersion: batch/v1
kind: CronJob
metadata:
  name: frigate-archive
  namespace: frigate
spec:
  concurrencyPolicy: Forbid
  schedule: "12 4 * * *"
  successfulJobsHistoryLimit: 1
  failedJobsHistoryLimit: 1
  jobTemplate:
    spec:
      backoffLimit: 0
      # 6 hours deadline
      activeDeadlineSeconds: 21600
      template:
        spec:
          restartPolicy: OnFailure
          containers:
          - name: archive
            image: genunix/sqlite:latest
            imagePullPolicy: Always
            command:
              - /bin/bash
              - -e
              - /config/archive.sh
            volumeMounts:
              - name: frigate-data
                mountPath: /media/frigate
              - name: frigate-archive
                mountPath: /media/archive
              - name: config
                mountPath: /config
          volumes:
            - configMap:
                name: frigate-config
              name: config
            - name: frigate-data
              persistentVolumeClaim:
                claimName: frigate-data
            - name: frigate-archive
              persistentVolumeClaim:
                claimName: frigate-archive
```

**Comment by toxic0berliner** [comment]
I also scripted something else based on unionfs and the POSIX creation date of the files to move them from hot to cold storage. I find this less elegant than the given script even if I believe mine can be run hourly instead of daily, but relies on unionfs...

Most important: we are several with the same need and strong enough to justify tweaking stuff on our side...

Would a dev maybe reconsider and take this topic? A few extra settings like archive path and time_before_archival, a scheduled job like the one above, and bingo ?

**Comment by NickM-27** [comment]
> Would a dev maybe reconsider and take this topic?

what do you mean reconsider? this issue is open and pinned, meaning it will be done at some point. It is however not a priority right now

**Comment by toxic0berliner** [comment]
My bad, I thought I recalled heavy scepticism in the beginning maybe causing actual devs or maintainers to loose interest. but my memory might be disserving me. Can fully understand priorities srry

**Comment by alsobrsp** [comment]
@fpytloun This looks perfect for my use case. I am currently writing everything to NFS.  Most of the other suggested solutions seem to be targeting local storage. I want the last day or so to be local with the rest remote.

**Comment by fpytloun** [comment]
Here is new version of my archivation script.

Changes:
- upload is now performed per file separately so process can be restarted at any time and also it can be run more frequently
- it supports `KEEP_DAYS=-1` which will migrate everything
- added retry function to bypass database locking. Would not be needed if frigate has option to use WAL journal type
- cleanup for empty leftover directories

```
    #!/bin/bash -e

    FRIGATE_BASEPATH="/media/frigate"
    ARCHIVE_BASEPATH="/media/archive"

    # Number of previous days to keep (-1 means copy everything)
    KEEP_DAYS=-1

    function retry {
      local retries=$1
      shift

      local count=0
      until "$@"; do
        exit=$?
        wait=$((2 ** $count))
        count=$(($count + 1))
        if [ $count -lt $retries ]; then
          echo "Retry $count/$retries exited $exit, retrying in $wait seconds..."
          sleep $wait
        else
          echo "Retry $count/$retries exited $exit, no more retries left."
          return $exit
        fi
      done
      return 0
    }

    for day in "${FRIGATE_BASEPATH}"/recordings/*; do
        day=$(basename "${day}")
        SKIP=0
        for range in $(seq 0 $KEEP_DAYS); do
            match=$(date -d "-${range} day" '+%Y-%m-%d')
            if [[ $day == "$match" ]]; then
                SKIP=1
                break
            fi
        done

        [[ $SKIP -ne 1 ]] || continue

        echo "[INFO] day=${day} Copying recordings into archive storage"
        find "${FRIGATE_BASEPATH}/recordings/${day}" -type f | while read recording; do
          target_file=${recording/$FRIGATE_BASEPATH/$ARCHIVE_BASEPATH}
          target_dir=$(dirname ${target_file})
          [ -d "${target_dir}" ] || mkdir -p "${target_dir}"

          echo "[INFO] day=${day}, file=${recording} Copying into ${target_file}"
          cp -a "${recording}" "${target_file}"

          echo "[INFO] day=${day}, file=${recording} Updating path in sqlite"
          echo "UPDATE recordings SET path = '${target_file}' WHERE path = '${recording}'" | retry 5 sqlite3 "${FRIGATE_BASEPATH}/frigate.db"

          echo "[INFO] day=${day}, file=${recording} Deleting original file"
          rm -f "${recording}"
        done
    done

    echo "[INFO] Cleaning up empty directories on archive storage"
    find "${ARCHIVE_BASEPATH}" -type d -empty -delete

    echo "[INFO] Cleaning up empty directories on hot storage"
    find "${ARCHIVE_BASEPATH}" -type d -empty -delete
```

**Comment by NickM-27** [comment]
> added retry function to bypass database locking. Would not be needed if frigate has option to use WAL journal type

Frigate does use the WAL journal type

**Comment by fpytloun** [comment]
> > added retry function to bypass database locking. Would not be needed if frigate has option to use WAL journal type
> 
> Frigate does use the WAL journal type

Iinteresting. Even when I used `PRAGMA journal_mode=WAL;` I sometimes hit database locked error.

**Comment by NickM-27** [comment]
Right, because there is still a write timeout that results in a locked error. If something is writing and something else waits for longer than that timeout it will have a locked error regardless of if it's using wal or not.

**Comment by pgiansante** [comment]
I tried the latest script above and I lose all access to the events via frigate app.
Does this work with the 0.13.2?

**Comment by fpytloun** [comment]
> I tried the latest script above and I lose all access to the events via frigate app. Does this work with the 0.13.2?

Yes, I am using it and it works fine.

**Comment by jakubsuchybio** [comment]
Hi guys, just want to report my findings with trying to use `unionfs`, then `mergefs`.
I managed to get it to work.
So basically this setup:
- local mount to 1TB SSD
- NFS mount to 12TB NAS
- union/merge fs mount which merged two above
this way the final storage said it has 13TB and writes from frigate would go defaultly to the first storage.
So the frigate thinks it saves to one storage, but they are actually 2 tiers.
Then I had a cron which would sync from local to NFS daily to clear up 300GB on the local storage. (My cameras takes like 150GB/day) so this would basically create a buffer for 2 days if something would go wrong at one day.

And go wrong it went constantly :(.
I don't have stable connection between my 2 locations (the NFS mount is not located on local storage, but across internet via VPN)
My network would sometime stall and disconnect.
That the unionfs and mergefs can't survive. They have some politics for disconnects, but I couldn't get them to survive disconnects.
Only way to restore the union/mergefs would be by restarting whole frigate vm :(

Having only archiving without being able to view the archive from frigate seems kinda pointless to me, that is why I was trying this solution.

One last solution that comes to my mind is to write custom script/program, which would basically do something similar to union/mergefs but with symlinks.
So when the local storage gets full, it will move oldest to the NFS archive and replace it locally with a symlink.
But that creates new problem. When the frigate tries to remove that by the retention, it will just remove the symlink and not the target of the link on the archive storage.
So to get it to work I would have to keep track of the symlinks on some db (file, sql, or something) and when going through, if some symlink would get deleted by frigate, but it was still tracked in the db, the script/program would delete it also from the NFS archive..

As I'm thinking about it, it might actually work. I will try to do that sometime soon and report back the findings.

**Comment by toxic0berliner** [comment]
I think this feature request is quite enough justified, I believe it's only waiting for some dev to start tackling it.
My opinion is archive will be slow and potentially subject to issues during transfer, but frigate could still let the user mount whatever archive storage as a separate path. Seems to me a more reasonable starting point than aiming at handling smb,NFS,S3,....
For me frigate should have 2 new settings: a keep time after which files are moved to the archive, and a path to said archive.
It should handle the copy and delete files from the main storage only once the archive has been successfully written. It should log any failure but also simply retry the transfer at the next occasion.
It should also handle actual deletion from the archive instead of the main storage when the usual keep policy expires.
Storage page could also benefit from being modified to display main and archive storage space.
I don't think it should be more complicated than that. (That said I also believe it's too much for me to try implementing so thanks in advance to anyone willing to tackle it)

**Comment by johntdyer** [comment]
>One last solution that comes to my mind is to write custom script/program, which would basically do something similar to union/mergefs but with symlinks.
So when the local storage gets full, it will move oldest to the NFS archive and replace it locally with a symlink.
But that creates new problem. When the frigate tries to remove that by the retention, it will just remove the symlink and not the target of the link on the archive storage.
So to get it to work I would have to keep track of the symlinks on some db (file, sql, or something) and when going through, if some symlink would get deleted by frigate, but it was still tracked in the db, the script/program would delete it also from the NFS archive..

@jakubsuchybio  - Would rsync w/ the delete flag work ?  Just thinking when Frigate deletes the linked file the rsync might recognize this ( untested ) and delete the linked file on the other storage...   Just an untested idea.... pre-coffee :)

**Comment by fpytloun** [comment]
@jakubsuchybio Frigate recordings has path in database so you can have multiple paths instead of using mergefs which seems to be very fragile. So why don't you use my script?
https://github.com/blakeblackshear/frigate/issues/3673#issuecomment-2057730850

**Comment by jakubsuchybio** [comment]
@fpytloun 

Juicefs, didn't hear of it, will look at it, thx for hint 

yeah i wouldn't dare to modify db paths 😁

**Comment by zzachattack2** [comment]
+1 for the feature request for tiered storage. I think the flexibility that Frigate already has on _what_ to record, and _how long_ to retain it, are great, but the missing part of _where_ is a big shortcoming IMO -- especially considering the volume of data that can be involved here.

Just throwing out some ideas, but what I think would be ideal is if we could at least identify a "hot-storage" recording path and a "cold-storage" recording path, and each recording 'option' that currently allows us to specify a retention period,  could also specify an archiving period before moving to cold storage. 

Example use case: 
I want to keep a week's worth 24/7 recordings (~4TB), which means I pretty much need to store that on spinning platters. For _recent_ footage of particular importance (e.g events with a person identified), which are substantially more likely to actually be viewed, it would be better to have it available on a faster storage medium. I also want to keep those events for much longer (say 30 days), but after a few days, the need for them on fast storage largely goes away (along with available space on the SSD), and they should be moved over to cold storage. 

So a hypothetical configuration could look like:
```
record:
  enabled: True
  hot_path: /data/hot
  cold_path: /data/cold
  retain:
    mode: all
    days: 7
    archive: 0 #don't use hot-storage
  events:
    retain:
      mode: active_objects
      default: 
        days: 14
        archive: 1 #move to cold-storage after 1 day
      objects:
        person:
          days: 30
          archive: 3 #move to cold-storage after 3 days
```

**Comment by javydekoning** [comment]
> +1 for the feature request for tiered storage. I think the flexibility that Frigate already has on _what_ to record, and _how long_ to retain it, are great, but the missing part of _where_ is a big shortcoming IMO -- especially considering the volume of data that can be involved here.
> 
> Just throwing out some ideas, but what I think would be ideal is if we could at least identify a "hot-storage" recording path and a "cold-storage" recording path, and each recording 'option' that currently allows us to specify a retention period, could also specify an archiving period before moving to cold storage.
> 
> Example use case: I want to keep a week's worth 24/7 recordings (~4TB), which means I pretty much need to store that on spinning platters. For _recent_ footage of particular importance (e.g events with a person identified), which are substantially more likely to actually be viewed, it would be better to have it available on a faster storage medium. I also want to keep those events for much longer (say 30 days), but after a few days, the need for them on fast storage largely goes away (along with available space on the SSD), and they should be moved over to cold storage.
> 
> So a hypothetical configuration could look like:
> 
> ```
> record:
>   enabled: True
>   hot_path: /data/hot
>   cold_path: /data/cold
>   retain:
>     mode: all
>     days: 7
>     archive: 0 #don't use hot-storage
>   events:
>     retain:
>       mode: active_objects
>       default: 
>         days: 14
>         archive: 1 #move to cold-storage after 1 day
>       objects:
>         person:
>           days: 30
>           archive: 3 #move to cold-storage after 3 days
> ```

I like this and suggest splitting the datastore to it's own object given the (future) possibility for remote/cloud storage:

```yaml
  datastore:
  - name: my-hot-volume
    type: hostPath
    config:
      path: /data/hot
  - name: my-cold-bucket
    type: s3
    config:
      bucket: s3://my-bucket
      prefix: /frigate
      storage_class: 'STANDARD'
```

**Comment by luckylinux** [comment]
Not sure what is wrong with `rsync`.

I agree that the `--delete` /and similar `--delete-after`. `--delete-before`, ...) Options are **MISLEADING**: they will REMOVE from the DESTINATION Directory if the File doesn't exist anymore in the SOURCE Directory ! Which it will not since we just moved it (Duh).

What **should** work (I just tried it now for moving from SSD to HDD for most of the Storage) is to use the `--remove-source-files` Flag.
Example (I did this under Fedora and I disabled the `-X` Flag since it was triggering many Errors due to SELinux apparently not working on a NFS mounted Share):
```
rsync -aAUvH --remove-source-files /home/podman/containers/data/frigate-server/ /mnt/frigate
```

Maybe also the `--prune-empty-dirs` could be useful.

Of course we can also agree that this is not perfect because it will just move EVERYTHING.

**Comment by jakubsuchybio** [comment]
> Not sure what is wrong with `rsync`.
> 
> I agree that the `--delete` (and similar `--delete-after`. `--delete-before`, ...) Options are **MISLEADING**: they will REMOVE from the DESTINATION Directory if the File doesn't exist anymore in the SOURCE Directory ! Which it will not since we just moved it (Duh).
> 
> What **should** work (I just tried it now for moving from SSD to HDD for most of the Storage) is to use the `--remove-source-files` Flag. Example (I did this under Fedora and I disabled the `-X` Flag since it was triggering many Errors due to SELinux apparently not working on a NFS mounted Share):
> 
> ```
> rsync -aAUvH --remove-source-files /home/podman/containers/data/frigate-server/ /mnt/frigate
> ```
> 
> Maybe also the `--prune-empty-dirs` could be useful.
> 
> Of course we can also agree that this is not perfect because it will just move EVERYTHING.

Ofc this would work for archival purposes.
But that is not what we wanted.
We wanted some solution with an ability to look at the archived videos through frigate UI. Moving it will basically remove it from UI. => Not useful.

**Comment by luckylinux** [comment]
> > Not sure what is wrong with `rsync`.
> > I agree that the `--delete` (and similar `--delete-after`. `--delete-before`, ...) Options are **MISLEADING**: they will REMOVE from the DESTINATION Directory if the File doesn't exist anymore in the SOURCE Directory ! Which it will not since we just moved it (Duh).
> > What **should** work (I just tried it now for moving from SSD to HDD for most of the Storage) is to use the `--remove-source-files` Flag. Example (I did this under Fedora and I disabled the `-X` Flag since it was triggering many Errors due to SELinux apparently not working on a NFS mounted Share):
> > ```
> > rsync -aAUvH --remove-source-files /home/podman/containers/data/frigate-server/ /mnt/frigate
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Maybe also the `--prune-empty-dirs` could be useful.
> > Of course we can also agree that this is not perfect because it will just move EVERYTHING.
> 
> Ofc this would work for archival purposes. But that is not what we wanted. We wanted some solution with an ability to look at the archived videos through frigate UI. Moving it will basically remove it from UI. => Not useful.

Sure, but I thought @fpytloun Script had the same Issue (move files between 2 Storages) and only UnionFS/FUSE (for those of us without UnRAID) were the only Options to "merge" those Storages together.

Or has @javydekoning Configuration been implemented in the latest Release already ?

**Comment by fvdpol** [comment]
if you want to keep access to the archived videos from the frigate UI I suggest to have a look at script written by @fpytloun. I find it works quite well and definitely an acceptable workaround until archiving / tiered storage becomes part of the frigate feature set. 
 (see https://github.com/blakeblackshear/frigate/issues/3673#issuecomment-2057730850)

**Comment by luckylinux** [comment]
Aaah alright. Yeah I remembered a few Days I saw that SQL Query (don't know why I didn't see it again today, I must be too tired :rofl:).

**Comment by jakubsuchybio** [comment]
> if you want to keep access to the archived videos from the frigate UI I suggest to have a look at script written by @fpytloun. I find it works quite well and definitely an acceptable workaround until archiving / tiered storage becomes part of the frigate feature set.
> 
> (see [#3673 (comment)](https://github.com/blakeblackshear/frigate/issues/3673#issuecomment-2057730850))
> 
> The script moves files (depending on the age) to another location, and then updates the frigate database to reference the new location. The resulting "move" is therefore transparent to frigate/ui. This also makes it robust in case the migration gets stalled/stopped mid-way.

Uff thanks for reminding. I totally missed that script and was waiting for that tiered/archival process from frigate. Will give it a try.

**Comment by iscofield** [comment]
Another person migrating from BI and was looking for a way to have the tiered storage. Another +1 from me for this to eventually come.

**Comment by cammurray** [comment]
Throwing my support against this one. +1

The issue for me is where the storage is located, my heavy disks are in the garage (they're noisy, butthey could be stolen) where my fast disks are in the house (because they're SSD and silent). I'd like to record to the fast disks and then after x days move them to the slow disk.

The scripted solution is fine, but it's a bit of a hack. Anything that is updating the database in parallel to how it's expected, is it at risk of being nuked in some kind of future update.

**Comment by magma1447** [comment]
> The issue for me is where the storage is located, my heavy disks are in the garage (they're noisy, butthey could be stolen) where my fast disks are in the house (because they're SSD and silent). I'd like to record to the fast disks and then after x days move them to the slow disk.
> 
> The scripted solution is fine, but it's a bit of a hack. Anything that is updating the database in parallel to how it's expected, is it at risk of being nuked in some kind of future update.

More or less exactly my case as well. I have just been following the thread until now since I respect that it's not (currently?) in the scope of the project. But for bigger installations it's a really nice feature.

**Comment by toxic0berliner** [comment]
I've been using unionfs and my custom archive script to periodically move files from SSD to HDD.
It works reliably and fine but I figured running unionfs in my CT (proxmox) is the reason I'm stuck doing cold backups and can't do snapshot backups...

I would love for someone to take that up. Meanwhile I will give a try to manipulation of the db and the file location...

**Comment by cammurray** [comment]
> I've been using unionfs and my custom archive script to periodically move files from SSD to HDD.
> It works reliably and fine but I figured running unionfs in my CT (proxmox) is the reason I'm stuck doing cold backups and can't do snapshot backups...

When using unionfs, are you having to modify the database? or does the path remain the same (just the underlying filesystem changes)?

**Comment by toxic0berliner** [comment]
with unionfs I have 
- /hot mounted on an ssd
- /cold mounted on an hdd
- /merged is the union of the 2
so I pass /merged to frigate, and I have a script in crontab that check on /hot for files older than 1 day and simply mv them from /hot to /cold

they remain all available on /merged and therefore no fussing around in the frigate DB or whatever, all remains available... but... this uses fuse and makes my snapshot backups fail :-(

so I believe someone found we can edit the DB to change the path of some files in the DB, I assume if we ensure to move the files and make them available on the new path it should all be fine too... A bit worried editing what seems to me an sqlite db that frigate is suceptible to open/edit in // whenever it pleases, but hey...

**Comment by pcislocked** [comment]
i did the same as the above, just with mergerfs, and wrote a simple bash script to move the files as necessary(as the first faster disk fills up).

**Comment by kooskaspers** [comment]
For those who make use of ZFS: has anyone tried adding an SSD as "[cache device](https://docs.oracle.com/cd/E36784_01/html/E36845/gnjlj.html)", to just speed up the read-requests ?

**Comment by dlasher** [comment]
+1 

This please.

break it into LIVE / ARCHIVE, let frigate, on a scheduled job, decide what needs to be trimmed or moved.

**Comment by MivraMe** [comment]
Yep +1 I am looking for that !

**Comment by dlasher** [comment]
> For those who make use of ZFS: has anyone tried adding an SSD as "[cache device](https://docs.oracle.com/cd/E36784_01/html/E36845/gnjlj.html)", to just speed up the read-requests ?

(READ CACHE) If you want to add L2ARC  (which costs RAM) more often you're better to add RAM. 

(WRITE CACHE + SYNC) Adding SLOG on the other hand, is a beautiful job for Intel P3700 PCI-E cards, great lifetime (TBW) and they're cheap on ebay.

**Comment by fvdpol** [comment]
> > For those who make use of ZFS: has anyone tried adding an SSD as "[cache device](https://docs.oracle.com/cd/E36784_01/html/E36845/gnjlj.html)", to just speed up the read-requests ?
> 
> (READ CACHE) If you want to add L2ARC (which costs RAM) more often you're better to add RAM.
> 
> (WRITE CACHE + SYNC) Adding SLOG on the other hand, is a beautiful job for Intel P3700 PCI-E cards, great lifetime (TBW) and they're cheap on ebay.

another option to look into is using a Metadata Special Device.

On my pool with 2x raidz2 devices of 6 mechanical hard drives each, I found adding an SSD for the metadata gave me a great performance boost, especially for searching/browsing through the filesystem/directories. 
Since loosing the metadata device would mean loosing the pool, I opted for a 3-way mirror of the metadata special device. That 3-way mirror gives besides the fault tolerance the additional benefit of faster reads/more iops. 

see: https://forum.level1techs.com/t/zfs-metadata-special-device-z/159954

In past I experimented with the L2ARC cache, but was not impressed with the results on my workload (and hardware)-- so I ditched it in favor of the metadata/small-block device.

**Comment by heffneil** [comment]
It would be great to have an archive option.  I am currently using mergefs with a 8tb NVME and a 22 gig spinner.  I will move anything older than 15 days to the spinner with its direct drive reference to keep most recent data available on the NVME for quicker access.  It would be nice to be able to archive the storage to my SAN but retain the ability to search for object and see timeline history.

**Comment by NickM-27** [comment]
>  I am currently using mergefs with a 8tb NVME and a 22 gig spinner. I will move anything older than 15 days to the spinner with its direct drive reference to keep most recent data available on the NVME for quicker access. It would be nice to be able to archive the storage to my SAN but retain the ability to search for object and see timeline history.

you should not lose access to the recordings if this setup is done correctly. This is how many of the existing setups work, including mine (uses fuse.fs as part of Unraid) and Frigate works regardless of if the recordings have been moved

**Comment by heffneil** [comment]
> > I am currently using mergefs with a 8tb NVME and a 22 gig spinner. I will move anything older than 15 days to the spinner with its direct drive reference to keep most recent data available on the NVME for quicker access. It would be nice to be able to archive the storage to my SAN but retain the ability to search for object and see timeline history.
> 
> you should not lose access to the recordings if this setup is done correctly. This is how many of the existing setups work, including mine (uses fuse.fs as part of Unraid) and Frigate works regardless of if the recordings have been moved

I understand and I am not losing data but I am suggesting that I have another point to move recording videos and snapshots if they are older than x day or the hard drive runs out of space.  Currently there isn't a way to "archive" that to larger, slower storage and retain the database capabilities?

**Comment by NickM-27** [comment]
> Currently there isn't a way to "archive" that to larger, slower storage and retain the database capabilities?

of course there is, which is my whole point. This is exactly how unraid works, and Frigate has no idea (nor does it care) if the recordings are on the SSDs or the HDDs

**Comment by heffneil** [comment]
> > Currently there isn't a way to "archive" that to larger, slower storage and retain the database capabilities?
> 
> of course there is, which is my whole point. This is exactly how my unraid setup works, and Frigate has no idea (nor does it care) if the recordings are on the SSDs or the HDDs

Im confused.  I don't want to pool my SAN as part of the same virtual drive.  It seems less than ideal or reliable to do this.  How can I use remote storage that frigate would know was moved to another location?

**Comment by NickM-27** [comment]
>  It seems less than ideal or reliable to do this

not really, like I said this is how Unraid works. I have been running this setup for ~4 years now with Frigate and Jellyfin, without any issues. 

![Image](https://github.com/user-attachments/assets/9a77c3d3-f8f7-4088-bc2a-071dadf5c8a8)

> How can I use remote storage that frigate would know was moved to another location?

currently the best way is that Frigate won't know. It should appear as a single directory to Frigate, even though it is actually split.

**Comment by pcislocked** [comment]
my current/new workaround is to use a separate frigate instance that only records substreams 24/7 to separate hdd. main instance only records motion for 7 days and clips for 180

**Comment by slackr31337** [comment]
+1 

I think there should be an archive path that has a separate database and mount point.

Current storage 24x7
/media/frigate/recordings

Archive storage nfs or remote cloud FS 
/media/frigate/archive

Run archive job every x hours, etc

In the web GUI, the ability to browse and search archived recordings, snapshots, etc

**Comment by jakubsuchybio** [comment]
Thanks for the thorough explanation.

I was one of those, who was dealing with the remote network instabilities.
However now I moved to the location when most of my cameras are and I must say, the problem didn't go away. :)
On the contrary, the problem even expanded. Before it was only network instability issues.
I'm running frigate on the same server where I have my TrueNas now. (promox, 1st VM is Truenas, 2nd VM is Frigate)

One new issue I found out was just simple rebooting of the whole server. Frigate started before TrueNas and there is the new problem. I didn't want any hard-coded delays in proxmox. I tried dealing with this on the docker compose level -> it turns out, it is actually impossible doing that, because mounting NFS is before anything else in docker compose, so when NFS mounting failed, there was literally no way to retry it. -> It forced me to this solution https://github.com/jakubsuchybio/frigate-nfs-wrapper/ where I had to wrap the frigate docker in a custom startup script, which first tries to mount the NFS and only after the correct mount, it will start the frigate.

But there is another issue. What if my Truenas changes state to degraded or crashes or is restarted? That also can happen and that also breaks the NFS link to the frigate. (I haven't resolved this yet, before when remote I was trying some NFS configurations for retrying the mount, when it crashes, but couldn't find any options which would work). So I guess I will have to code that myself as well. Having some cron/daemon which will be checking the mount and if it fails, it should stop frigate and start it when NFS mounts.

I see the archive option in frigate much more usefull, because it will solve all these classes of problems. Because the archival is just copying/moving to some other storage, which doesn't need to be up 24/7 for the frigate to work. 

And I don't think that this functionality is so hard to implement. If I knew some python I would gladly code it myself into the frigate (if maintainer would allow merging it), but right now I'm learning Rust and learning another lang is not in my capacity right now :(

From my understanding only few things would need to be implemented.
1. Some configuration for connecting to the archive, there could be couple providers s3, nfs, sftp/ssh, smb
2. Other configuration for setting up rules for archival (size, time, space left, etc)
3. Periodic archival with graceful error handling and with replacing the path in the db
4. Continuously checking if the archival is speedy enough, meaning the connectivity is faster than recording speed of all cameras.

**Comment by venix1** [comment]
Great response, it does show the nature of adding it to frigate. What appears simple grows quickly once you start factoring in all sorts of environment and conditions, and once it's in the source.  It's expected to work. 

> But there is another issue. What if my Truenas changes state to degraded or crashes or is restarted? That also can happen and that also breaks the NFS link to the frigate.

I did assume a stable 24/7 connection in my write-up because it's what I demand from it, but NFS does crash on me occasionally and if the server is down, I do not want  that causing my cameras to go offline.  Full disclosure, I'm not using a network storage for this.  I'm content will dedicating HDDs  on a dedicated CPU.  The amount of work I expect from frigate is quite high, so dedicated hardware just makes sense.  I did the initial setup in VM with NAS environment.  So I understand that is a use case.   

Back to the point MergerFS fails just as you describe, because of NFS.  (Easily tested by blocking NFS traffic).  This is not a problem of mergerfs, or any filesystem union/fuse/native but how the kernel's vfs layer behaves.  If the filesystem is inaccessible the default logic is to hang IO operations indefinitely to prevent corruption.  In the event the NFS mount goes down, frigate would no longer be able to read or write. This is true of frigate itself, even if it's only doing archiving on NFS, attempting to read, or migrate data, would hang.  However, with timeouts and state tracking it could prohibit those operations until it detects NFS is back. So the application can continue working.

Maybe NFS doesn't have to hang.  This is going in the area of I don't know how well things work out, but there are options to mount NFS and fail quickly.  
```
mount 192.168.1.1:/nfs/frigate/ /data/remote -o soft,timeo=1,retrans=1
soft - Do not hang IO calls, allow them to timeout
timeo=1 - Timeout after 1 second
retrans=1 - Do not retry
```
The result, operations take a minimum of 1 second, not sure how frigate will handle that. And there may be better tweaks to mergerfs.  For instance, why does the FF policy hang?  If it's only writing to `/data/fast` why does it hang if `/data/remote` is down.  An option to ignore a down branch doesn't seem so bad to me either but it's specific to this use case and may not be desirable on that side. The only thing found related to mergerfs and NFS is a [5 year old post ](https://github.com/trapexit/mergerfs/issues/691) on the topic. 

After considering it, while I like mergerfs, given issues with remote filesystems, a more robust solution may actually be symlinks, as you've mentioned before,  created and managed by a dedicated script.
1. Frigate only operates on /data/fast 
1. Archive script will move files leaving a symlink pointing to the correct location (basically emulating mergerfs) 
1. If NFS is down, the archive script hangs.  (no big deal)
1. if NFS is down, reading from NFS archives is broken. (hopefully frigate will timeout to avoid resource exhaustion, but at least writes are unaffected, maybe still use `soft`/`timeo`)
1. More or less fully replaces mergerfs. You could use them together as long as `soft`/`timeo` is used. But it's not necessary at all. It probably adds little to no value at this point tbh. 
1. Zero fuse performance overhead in reads and writes  
1. Still needs custom free space logic, and deleted symlink pruning. 
   - rsync can work here, by having it show show which files would be copied from archive to fast storage and then delete those files.  Having more than 2 tiers of storage breaks this method. (It just gets overly complicated)  
   - It's simpler to just read the database, and prune things that don't exist. 
1. There may be a script out there that does this for a similar workload. 

There was mention of writing paths to database in lieu of symlinks.  It works if frigate shares write access to it, and it might have issues with NFS outages still. Leaning on the side of caution, reading it is fine, but modifying I'd avoid.  Assuming there isn't an issue with symlinks and frigate that I missed.  

It should be noted most of the NFS issues and theoretical what-ifs are common problems with NFS/remote filesystems in general. A single NFS host is terrible for reliability(production setups are redundant for a reason).  .  It's great for home use, but frigate which should run 24/7 is actually outside of that design paradigm and closer to a business/enterprise solution.  A solution for NFS not being ready when frigate starts has too many variables. Waiting for NFS is the best course of action, but how is entirely dependent upon the setup.

**Comment by toxic0berliner** [comment]
I think we should keep this on track, every other NVR implements something for archival or tiered storage, I can't imagine frigate will indefinitely leave us using esoteric filesystems or manually digging in it's database to find a solution.
Hope a dev can implement the DB trick above in a cleaner way at least as a starting step.
Other considerations like the best NFS settings to store frigate recordings should be another discussion I believe.
Not sure it's even worth discussing other workarounds one might have found to fill this missing feature, I hope this turns into how to best implement it fast, simple and maintainable.

**Comment by dlasher** [comment]
> I think we should keep this on track, every other NVR implements something for archival or tiered storage, I can't imagine frigate will indefinitely leave us using esoteric filesystems or manually digging in it's database to find a solution. Hope a dev can implement the DB trick above in a cleaner way at least as a starting step. Other considerations like the best NFS settings to store frigate recordings should be another discussion I believe. Not sure it's even worth discussing other workarounds one might have found to fill this missing feature, I hope this turns into how to best implement it fast, simple and maintainable.

I completely agree. Since Frigate is already taking care of housekeeping (staying within Quota, expiring old, etc) it makes far more sense to move the logic for archiving to Frigate itself.

+1 from me.

**Comment by NickM-27** [comment]
This feature request is pinned, which means we intend to implement it at some point. There's no need to continue to advocate for the feature.

**Comment by platima** [comment]
> "Hot" mounted at: /media/frigate/recordings/hot
> "Cold" mounted at: /media/frigate/recordings/cold

@javydekoning I completely agree - I want this. I'd have my local SSD as the hot storage that by default everything is recorded to, and an NFS export as the cold storage.

The only thing I would perhaps suggest is that it doesn't break the current hierarchy too much, eg:
- Hot mounted at: /media/frigate
- Cold mounted at: /media/frigate_archive

This is essentially what I'm doing now, and just manually moving files then doing 
```
record:
  sync_recordings: True
```

**Comment by smitty-nieto** [comment]
I too would like to see something around this or being able to re-scan and pick up directories that may have been restored.

Example : We have archive/long term storage of days in Backblaze. The client, an upscale restaurant  gets a lawsuit that for someone that fell 3 months ago and we need to restore that specific day.

How can we handle this or allow us to point to the offsite storage (mounted locally via rclone).

**Comment by pygmymarmoset** [comment]
@fpytloun Are you still using this same code?  Are you running frigate in docker?  It seems to be from an older (?) setup, at least if using docker, where config and mount weren't separate?  

Wondering because cron isn't running inside the container (and sqlite3 client isn't installed), but if it's done external to the container I think you risk corrupting the sqlite db because locking isn't always handled across the docker boundary. 

Thanks for any information about your setup!

**Comment by pbix** [comment]
@fpytloun.  Your script solutions look great but for those less container literate could you elaborate just a little more on where this script is located?  I run Frigate in an lxc under Proxmox.  When I open the shell and type "sqlite3" I get "command not found".  Makes me think I cannot run it in my lxc container.

**Comment by luckylinux** [comment]
@pbix: for Podman/Docker you typically bind-mount a File/Folder from the Host to the Container, in the location you want. Not sure if there is any specific Place it should be bound to/from.

Eample for Podman (Quadlet):
```
Volume=/home/podman/containers/scripts/frigate/server/archive_storage.sh:/opt/archive_storage.sh:ro,z
```

Example for Docker (Compose):
```
volumes:
            - /home/podman/containers/scripts/frigate/server/archive_storage.sh:/opt/archive_storage.sh:ro,z
```

Enter the Container with:
```
podman exec -it frigate-server /bin/bash
```

Or execute a Command directly with (e.g. via Systemd Timer / CRON):
```
podman exec frigate-server my-command
```

(substitude `podman` with `docker` if using Docker instead of Podman)

Note that Files that you create inside the Container and are **not** bind-mounted to an External Volumes will be LOST at subsequent Container/Host Restart !

**Comment by kravemir** [comment]
> We don't want frigate managing the native filesystems, we want one volume mount for recordings without caring what the actual file system is behind that.

Technically, Frigate NVR manages recordings in storage(s), not filesystem. It already does for just one storage.

Filesystem doesn't know real usage pattern of files. Neither it knows the structure of directories and files application uses. Neither it knows metadata and rules (what camera, what retention policy, in what storage,...).

This is application level responsibility. One implementation is S3 objects, which provides API.

Another implementation is scripting and gluing things together with MergerFS. And, it needs to be maintained in each deployment. If something changed in the application's file structure, then script needs to be updated in all deployments. It is not really something to DIY for each deployment individually.

It is implemented by:

- [Viseron](https://viseron.netlify.app/components-explorer/components/storage/#advanced-configuration) - Storage Tiers.
- [iSpy](https://www.ispyconnect.com/docs/agent/storage) - Media Archive - ideal for SSD and HDD (archive) split.
- [ZoneMinder](https://zoneminder.readthedocs.io/en/latest/userguide/filterevents.html) - filtering with Move all matches action.
- [Kerberos.io](https://kerberos.io/) - separates Agent (record), Vault (store) and Hub (process/aggregate/view) composed separately.

TBH:

- I'm considering to migrate to Viseron just because of this feature. If Viseron had object-tracking feature, then I had already decided. I'll evaluate Viseron anyways.
- Kerberos.io seems great, but only Agent is open-source, Vault is free closed-source and Hub is paid (more than 5 cameras) closed-source.

**Comment by toxic0berliner** [comment]
I disagree with the tone of the message but not the facts. 
Migrate away all you want threats are not welcome here you are using FOSS deal with it. 

That being said it is indeed application resposability to manage more than one storage with internal rules and let users map storage backends to whatever is supported. I do agree with that.

I hope this can be taken up by a dev. 

And having opened the issue I do have the close button I will not let anyone be harassing the team, be nice in this issue or I'll close it and you can open your own to see if being mean helps you get what you want for free somewhere else.

**Comment by fpytloun** [comment]
@kravemir what about grabbing Claude Code and implementing storage tiers in Frigate? If this is only thing you miss in Frigate then it might be still highest ROI than migrating to something else unless it is better in other areas too.

I am still using my original script with Frigate but I will be refreshing my NVR setup in few months so maybe I will jump into considering the same thing.

**Comment by kravemir** [comment]
> I disagree with the tone of the message but not the facts. Migrate away all you want threats are not welcome here you are using FOSS deal with it.

@toxic0berliner I disagree with my personal choice being a threat. Who does it threaten and with what? 🤷‍♂️

**Comment by toxic0berliner** [comment]
dude, this is not the place to explain that you want out of here because other projects are better and compare pricing and strategies.
Focus : archive storage is the topic.
Comparing how others do this is ok to help the argument that it needs to be handeled in some way by frigate not our own workaround at filesystem level.
Comparing pricing licences and explaining you will migrate away is definitively not the topic and can only serve to annoy the devs in claiming their software is subpar which it isn't.

Lastly a bit of warning, I don't know the AI stance of frigate project, I just know you'll need patience as changes in storage run deep and you I do'nt expect you can create a single merge request with tousands of new lines of code that an AI spat out for you on such a mature project.
I see this as several changes that are all retro-compatible to use storage in a different way in all places, one by one, until we can create a storage manager efficiently. Just my 2 cents of thought.

**Comment by kravemir** [comment]
> dude, this is not the place to explain that you want out of here because other projects are better and compare pricing and strategies.
> Focus : archive storage is the topic.

@toxic0berliner Archive storage is the topic.

I highlighted about this topic, that other projects implement it and that it's important feature to me, that it affects my decision making, And it's not just a nice to have thing for me.

Personal importance of features for people is an useful information for maintainers. It helps sorting out and prioritizing things.

If there's just one person (e.g. me) strongly feeling about a feature, then probably it's not important. If many people consider it as "deal breaker", that it's worth considering as a priority.

If you have constructive feedback how to share personal importance of features in a better tone, I'll welcome it.

**Comment by toxic0berliner** [comment]
there's over 100 comments in this issue already, I don' think we need one more "I'll switch away from frigate" to change anyone's mind...
Again, be mindfull of how others and especially contributors to this project might recieve your feedback.

Let's leave the meta-talk aside and focus on how we need this feature to work and how it can best be implemented, but I agree that without contributors in the talk it's a bit vain for now...

Anyone able to start digging in the code and see what could be done ?

**Comment by NickM-27** [comment]
This feature request is marked as planned and we are well aware of it, but until multi stream recording (which is in progress) is implemented this won't make sense to implement.
